### PR TITLE
Hierarchical config support for targets

### PR DESCRIPTION
**Pull request by luigiberrettini**
I simply added the _SetItemFromElement_ method in _XmlLoggingConfiguration_ so that _SetPropertyFromElement_ calls it when parsing the configuration
I then added a test and the needed fake Target.

**NLog version**
4.3.0

**Work arounds**
None

**Config when implemented**
Take a look at [my Syslog target fork] (https://github.com/luigiberrettini/NLog.Targets.Syslog/blob/master/src/TestApp/NLog.config)
The GitHub version of the target does not work because it references the official NuGet package of NLog 4.3.0 (I tested it locally with the changes to NLog included in this PR)
```xml
    <target xsi:type="syslog:Syslog" name="syslog5424-tgt">
      <syslog:layout xsi:type="SimpleLayout" text="${message} ${exception:format=ToString,StackTrace}" />
      <syslog:sysLogServer>127.0.0.1</syslog:sysLogServer>
      <syslog:port>1514</syslog:port>
      <syslog:facility>Local4</syslog:facility>
      <syslog:protocol>UDP</syslog:protocol>
      <syslog:ssl>false</syslog:ssl>
      <syslog:splitNewlines>false</syslog:splitNewlines>
      <syslog:rfc>Rfc5424</syslog:rfc>
      <syslog:rfc3164 hostname="notUsedHostname" tag="notUsedTag" />
      <syslog:rfc5424 protocolVersion="1" hostname="${machinename}" appName="${logger}" procId="${processid}" msgId="testMsgId">
        <syslog:structuredData>
          <syslog:sdElement sdId="thread@${enterpriseId}">
            <syslog:sdParam name="id" value="${threadid}" />
            <syslog:sdParam name="myParam" value="myValue" />
          </syslog:sdElement>
          <syslog:sdElement sdId="timeQuality">
            <syslog:sdParam name="tzKnown" value="1" />
            <syslog:sdParam name="isSynced" value="1" />
            <syslog:sdParam name="syncAccuracy" value="60000000" />
          </syslog:sdElement>
          <syslog:sdElement sdId="origin">
            <syslog:sdParam name="ip" value="10.10.10.1" />
            <syslog:sdParam name="ip" value="192.0.2.1" />
            <syslog:sdParam name="enterpriseId" value="${enterpriseId}" />
            <syslog:sdParam name="software" value="MyAmazingSw" />
            <syslog:sdParam name="swVersion" value="3.5" />
          </syslog:sdElement>
          <syslog:sdElement sdId="meta">
            <syslog:sdParam name="sequenceId" value="${sequenceId}" />
            <syslog:sdParam name="sysUpTime" value="30200" />
            <syslog:sdParam name="language" value="en-US" />
          </syslog:sdElement>
        </syslog:structuredData>
      </syslog:rfc5424>
    </target>
```

### CODE REVIEW

**Comment by luigiberrettini** [comment]
The reference to _System.DirectoryServices.AccountManagement_ has been committed:
https://github.com/luigiberrettini/NLog/commit/325d3163caab371ed00f21c2acedafa30bfcbb53

Do you know how I can pass the builds?

**Comment by 304NotModified** [comment]
Hi! Thanks for this PR. 

But I need some context. What are we trying to solve here? Why do we need the StructuredConfigDebug target and those helper classes in NLog?

**Comment by luigiberrettini** [comment]
Hi @304NotModified, as you can see from the above XML, I would like to be able to parse configuration nodes that are not a layout and do not contain simply text, i.e. XML nodes with attributes or children.
The test needed a target with such a structure and I did not want to dirty the Debug target: I can make the StructuredConfigDebug internal and the classes it is using if you prefer.

**Comment by 304NotModified** [comment]
I think StructuredConfigDebug target etc should be in the user code?

**Comment by luigiberrettini** [comment]
Yes near the tests: I looked at similar tests in the same class and did he same as the Debug target disconnecting the brain a second before doing it :)
I will put he classes in the same file of the tests if this is ok for you and push it.
If you think the PR is ok after these changes can you already do a rough estimate on the time needed to have a new NuGet version?

**Comment by luigiberrettini** [comment]
I put he classes in the same file of the tests.

If you think the PR is ok after these changes can you already give me a rough estimate of the time needed to have a new NuGet version?

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 1]
```diff
@@ -1139,6 +1127,42 @@ private bool SetLayoutFromElement(object o, NLogXmlElement layoutElement)
...
+                return false;
+            }
+
+            object item = propInfo.GetValue(o, null);
+            this.ConfigureObjectFromAttributes(item, element, true);
+            this.ConfigureObjectFromElement(item, element);
+            return true;
+        }
+
+        private void ConfigureObjectFromAttributes(object targetObject, NLogXmlElement element, bool ignoreType)
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: Please undo the move of the method(s) . We need to merge this also with the coreclr branch, and now it's likely to get conflicts

**Comment by 304NotModified** [comment]
PS thinking to release this before 1 may, but I need to check the possible impact of the change first.

**Comment by 304NotModified** [comment]
And we needs docs before releasing it. (with examples etc) .

**Comment by 304NotModified** [comment]
I think we can improve the example in the unit tests, as I now find it a bit confusing. Do you agree on that? 

Also, I think we should rename this change to something like : "support for configuring target parameters with xml fragments" . What do you think?

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 1 above)
Comment (luigiberrettini) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: I will do it by Wednesday

**Comment by luigiberrettini** [comment]
It was hard to find a good commit message: whatever you think is more suitable is ok for me.

A more meaningful example can be done if I mimic what I did in [my Syslog target fork](https://github.com/luigiberrettini/NLog.Targets.Syslog/tree/master/src/NLog.Targets.Syslog) with the SyslogTarget, Rfc5424, StructuredData, SdElement and SdParam classes.

If you need me to do something on documentation or tests please let me know and I will try to do my best to help.

**Comment by luigiberrettini** [comment]
I do not understand why AppVeyor cannot find the CreateTestUsers.cmd that are still present in this branch :(

**Comment by 304NotModified** [comment]
testing this right now

**Comment by 304NotModified** [comment]
There was something strange with the last merge commit. I rebased the PR and changed the example. #1401 supersedes this PR.

**Comment by 304NotModified** [comment]
@luigiberrettini 

This has been released in 4.3.2 and is now online! https://www.nuget.org/packages/NLog/4.3.2

**Comment by luigiberrettini** [comment]
Thank you so much!
:smile:

**Pull request closed without merging**